File: packages/typescript/src/ubigeo/searchDistricts.ts
import { districtList } from "./data";
import { normalize } from "./normalize";
import type { UbigeoRecord } from "../types";

function rank(name: string, query: string): number {
  if (name === query) {
    return 0;
  }
  if (name.startsWith(query)) {
    return 1;
  }
  if (name.includes(" " + query)) {
    return 2;
  }
  if (name.includes(query)) {
    return 3;
  }
  return -1;
}

export function searchDistricts(query: string, limit = 10): UbigeoRecord[] {
  const normalized = normalize(query);
  if (normalized.length === 0) {
    return [];
  }
  const scored: { record: UbigeoRecord; score: number; index: number }[] = [];
  districtList.forEach((record, index) => {
    const score = rank(normalize(record.name), normalized);
    if (score >= 0) {
      scored.push({ record, score, index });
    }
  });
  scored.sort((a, b) => a.score - b.score || a.index - b.index);
  return scored.slice(0, limit).map((item) => item.record);
}

File: packages/rust/src/ubigeo/search_districts.rs
use super::record::UbigeoRecord;
use crate::data::ubigeo::DISTRICTS;

pub fn search_districts(query: &str, limit: usize) -> Vec<UbigeoRecord> {
    let q = normalize(query);
    if q.is_empty() {
        return Vec::new();
    }

    let mut scored: Vec<(u8, usize, &crate::data::ubigeo::Row)> = Vec::new();
    for (idx, row) in DISTRICTS.iter().enumerate() {
        let name = normalize(row.1);
        if let Some(rank) = rank_match(&name, &q) {
            scored.push((rank, idx, row));
        }
    }

    scored.sort_by(|a, b| a.0.cmp(&b.0).then(a.1.cmp(&b.1)));
    scored
        .into_iter()
        .take(limit)
        .map(|(_, _, row)| UbigeoRecord::from_row(row))
        .collect()
}

fn rank_match(name: &str, q: &str) -> Option<u8> {
    if name == q {
        return Some(0);
    }
    if name.starts_with(q) {
        return Some(1);
    }
    if is_word_start(name, q) {
        return Some(2);
    }
    if name.contains(q) {
        return Some(3);
    }
    None
}

fn is_word_start(name: &str, q: &str) -> bool {
    name.split(|c: char| c == ' ' || c == '-')
        .any(|word| word.starts_with(q))
}

pub(crate) fn normalize(input: &str) -> String {
    input
        .trim()
        .chars()
        .map(strip_accent)
        .collect::<String>()
        .to_lowercase()
}

fn strip_accent(c: char) -> char {
    match c {
        'á' | 'à' | 'ä' | 'â' | 'ã' | 'Á' | 'À' | 'Ä' | 'Â' | 'Ã' => 'a',
        'é' | 'è' | 'ë' | 'ê' | 'É' | 'È' | 'Ë' | 'Ê' => 'e',
        'í' | 'ì' | 'ï' | 'î' | 'Í' | 'Ì' | 'Ï' | 'Î' => 'i',
        'ó' | 'ò' | 'ö' | 'ô' | 'õ' | 'Ó' | 'Ò' | 'Ö' | 'Ô' | 'Õ' => 'o',
        'ú' | 'ù' | 'ü' | 'û' | 'Ú' | 'Ù' | 'Ü' | 'Û' => 'u',
        'ñ' | 'Ñ' => 'n',
        'ç' | 'Ç' => 'c',
        other => other,
    }
}
